File: missing-number/obzva.cpp
/**
 * 풀이
 * - 특정 정수가 `nums` 배열을 통해 주어졌는지 여부를 체크하는 배열 `check`를 만듭니다
 * - `nums`를 조회하며 `check`배열의 값들을 변경합니다
 * - `check`배열을 조회하여 누락되었던 정수를 확인합니다
 * 
 * Big-O
 * - N: 주어진 배열 `nums`의 크기
 * 
 * - Time Complexity: O(N)
 *   - 배열 `nums`를 조회하는 반복문은 O(N)의 시간 복잡도를 가집니다
 *   - 배열 `check`를 조회하는 반복문 또한 O(N)의 시간 복잡도를 가집니다
 *   - 따라서 전체 시간 복잡도는 O(N)입니다
 * 
 * - Space Complexity: O(N)
 *   - `check`배열의 크기가 입력값에 비례하여 선형적으로 증가합니다
 */

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int n = nums.size();
        
        vector<bool> check(n + 1, false);

        for (auto num : nums) check[num] = true;

        int res;
        for (int i = 0; i < n + 1; i++) {
            if (!check[i]) {
                res = i;
                break;
            }
        }

        return res;
    }
};

File: find-median-from-data-stream/obzva.go
/*
풀이
- 이진탐색을 이용합니다
Big O
- N: 현재 MedianFinder.nums의 크기
- AddNum
  - Time complexity: O(N)
    - bisect -> O(logN)
	- slices.Insert -> O(N)
  - Space complexity: O(1)
- FindMedian
  - Time complexity: O(1)
  - Space complexity: O(1)
*/

import "slices"

type MedianFinder struct {
	nums []int
}

func Constructor() MedianFinder {
	mf := MedianFinder{}
	mf.nums = make([]int, 0)
	return mf
}

func (this *MedianFinder) AddNum(num int) {
	n := len(this.nums)
	if n == 0 {
		this.nums = append(this.nums, num)
	} else {
		idx := bisectLeft(this.nums, num)
		this.nums = slices.Insert(this.nums, idx, num)
	}
}

func (this *MedianFinder) FindMedian() float64 {
	n := len(this.nums)
	if n%2 == 0 {
		return (float64(this.nums[n/2-1]) + float64(this.nums[n/2])) / 2
	} else {
		return float64(this.nums[n/2])
	}
}

// ----- Helper -----
func bisectLeft(arr []int, x int) int {
	lo := 0
	hi := len(arr)
	for lo < hi {
		mid := lo + (hi-lo)/2
		if arr[mid] < x {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddNum(num);
 * param_2 := obj.FindMedian();
 */
